src/pytezos/crypto/key.py

```diff
@@ -149,7 +158,7 @@ def from_secret_exponent(
             pk = fastecdsa.keys.get_public_key(bytes_to_int(secret_exponent), curve=fastecdsa.curve.P256)
             public_point = fastecdsa.encoding.sec1.SEC1Encoder.encode_public_key(pk)
         else:
-            assert False
+            raise AssertionError()
```

Comment:
Err.. this is not going to work in a prod env.

AssertionErrors are a warning, not an error.

Let's hope that this does not turn out to be important.

Reply:
oh what do you mean? Won't it result in a 500? How would they get suppressed to a warning?

Reply:
So exceptions are split into two types, warnings and errors.

But AssertionError is a weird one because it is equivalent to writing `assert False` (which always trips).

It's implementation, however, relies on the `__debug__` flag being set, which only occurs when you run Python in non-optimised mode (i.e. in a unit-test):
https://docs.python.org/3/reference/simple_stmts.html#assert

In production environments, like in C, it is very rare to run unoptimised code, and so asserts will never trigger on (say) a lambda.

This means, that the only way to pick up on code that is failing an assert is to run it in a debugger, and then you get to see all of the nasties.

However, if these asserts are fixes against errors within their library (only) then it is fine.

E.g.
```
internal_param = True
...
def foo(user_param):
   assert internal_param, "This must be true"  # This is fine because internal_param is a self-check only
   assert user_param, "User did not set user_param" # This is an anti-pattern because it will only trigger in a unit-test, instead a ValueError (or similar) should be raised
```

In my dumb example above, assert (or raise AssertionError) is fine if you are checking your own library, but it should not be relied upon to check the callers syntax.

It is not clear to me whether they are using the assertion correctly, but either way, the code behaviour has changed considerably.

Before:
```
  def tezos_fn(...):
      ...
      else:
          return False, "some reason"
```

After
```
  def tezos_fn(...):
      ...
      else:
          if __debug__:
              raise Exception("some reason...")
          return None  # May result in an exception due to a type mismatch
```

If they had replaced the code with a ValueError, then this would be fine as the error would be clear to both library and caller, but now, in production, the risk is that we will either get a silent failure message (code carries on running) or it will generate weird (and therefore hard to debug) faults such as type mismatches in our code (when the fault was actually in pytezos due to a bad user param)

Reply:
So short of us "trusting" their code, or running or prod in `__debug__` mode, I cannot see how we are going to spot potential issues, unless we run everything in a debugger/unit-test first. 😢

Reply:
btw - I don't think this is a major problem.

The fact that they have changed the API (tuple -> None) means that our backend is likely to trip up quite spectacularly if we do hit one of these changed lines, and so will be easy to find (even in production).

But we will have lost any hope of seeing the reason for the problem, making it harder to debug.

Reply:
thanks for the explanation, I wasn't aware that the `__debug__` flag could do this. The idea of letting errors be suppressed when we most need our code to be valid, is not the way I would have done it :/

Reply:
It might be right if they are only trying to catch their own bugs, but seeing how they are mis-exceptions elsewhere, my guess is that they don't.